[자동차 경주 구현] 악어 미션 제출합니다.

Hogu59 opened the pull request:
안녕하세요, 페어 4조 악어입니다!
많이 부족한 코드인데, 가르쳐 주시면 열심히 배우도록 하겠습니다!

❓질문 (미션을 진행하면서 궁금한 점이 있어 질문 남깁니다!)
 1)  입력값 검증(Validation)을 InputView 에서 하는게 좋은지, 아니면 model 혹은 controller 에서 진행하는 것이 좋은지 알고 싶습니다.
 (이번 미션에서는 model과 controller에서 유효성 검증을 했는데, depth가 깊어지는 문제가 있었습니다... 혹시 depth를 줄이는 방법이 있다면 알 수 있을까요..?)

 2) Race.kt
 - Controller로 Race class 를 사용하고 있습니다. 이때, getNames() 메서드에서 OutputView를 통해 입력가이드문을 출력하는 부분이 있는데, 이러한 출력부를 별도의 메서드로 분리하는 것이 좋은지, 아니면 메소드명을 수정하는 것이 좋은지 알고 싶습니다!
```kotlin
private fun getNames(): List<String> {
        outputView.printCarNamesGuide()
        lateinit var names: List<String>

        while (true) {
            try {
                names = inputView.readCarNames().split(COMMA)
                validateNames(names)
                names.forEach { name ->
                    Car(name)
                }
                break
            } catch (e: IllegalArgumentException) {
                println(e.message)
            }
        }
        return names
    }
```

감사합니다!

Hogu59 committed:
docs: Add requirement in README.md

Add fundamental requirements.
Add requirements for exception handling (Car, RoundNumber)

Hogu59 committed:
feat(view): Implement print car name input guide

view package created.
Printing Car name input Guide announcement function is added to OutputView.kt

Hogu59 committed:
docs: modify readme(outputview - car name request)

Readme file modified. OutputView -> Car name input Guide announcement printed correctly.

Hogu59 committed:
feat(view): implement car names input function

Function getting input of car names is added to InputView.
Return type is String, and the validation would be conducted in Controller and Model.

Hogu59 committed:
feat(view): implement print round number guide

Printing round number input guide announcement function added.

Hogu59 committed:
feat(view): implement input round number function

Function getting input of round number is added to InputView.
Return type is String, and the validation would be conducted in Controller and Model.

Hogu59 committed:
docs: modify Readme(car name, round number)

Readme file modified.
input and output functions related with car name and round number are functioning correctly.

Hogu59 committed:
fix(view): Add constant for round number guide

Constant for Round Number input announcement added.

Hogu59 committed:
feat(view): implement print result header

Function for printing execution result header added.

Hogu59 committed:
docs: modify readme(outputview - result header)

Readme file modified. OutputView -> result header printed correctly.

Hogu59 committed:
feat(model): implement car class

Car (data) class added for model.

Hogu59 committed:
feat(controller): implement get name list function

Controller package added.
Race class added.
getNames function, makes string to List<String> from inputView data, added.

Hogu59 committed:
feat(controller): implement validation nameList

Function for validating name list added.

Hogu59 committed:
feat(model): validate car name in model

Function for car name validation added.
Exceptions,
1) Empty input
2) name with space

Hogu59 committed:
feat(model): generate random number

Function for generating random number added.

Hogu59 committed:
feat(model): determine when to move

Function for determining when to move added.
If random number is same or greater than 4, the position would be changed)
(as, move was conducted)

Hogu59 committed:
feat(controller): implement getting round number

Function getting round number added.
Get string input from InputView, and return int.

Hogu59 committed:
feat(controller): validate round number

Function for validating round number input added.
Exceptions,
1) input is out of int range
2) input is not number (ex. letter, etc.)
3) input is empty

Hogu59 committed:
feat(controller): implement getting car list

Function returning List<Car> from List<string> (names) added.

Hogu59 committed:
docs: modify readme(determine move)

Function determining to move is working properly.

Hogu59 committed:
Feat(controller): execute all rounds

Start function added.
Function for executing all rounds added.

Hogu59 committed:
feat(view): print results for each round

Function to print current status(name, position) was added.

Hogu59 committed:
feat(view): print new line

Function to print new line added.

Hogu59 committed:
feat(controller): print results for all rounds

Implement printing current car status. (name, position)

Hogu59 committed:
docs: modify readme(print result for each round)

result for each round printed correctly.

Hogu59 committed:
feat(controller): get winner list

Function getting winner list added.
Get winner list, list of strings(name).

Hogu59 committed:
feat(view): print winners

Function for printing winner list added.

Hogu59 committed:
feat(controller): show winners

Function to print winner list added in Controller.

Hogu59 committed:
feat(controller): use functions for result

Function related with result printing called in start function.

Hogu59 committed:
feat(application): call start function in main

Call start function for Controller in main function of Application.

Hogu59 committed:
test: test for car name exceptions

Exception checked,
1) empty name input
2) name length 0 or more than 5
3) name with space

Hogu59 committed:
test: test for race class(controller)

Exception checked,
1) empty name list
2) size of name list is less than 2
3) size of name list is greater than 20
4) when duplicated name exist
5) when round number input exceeds int range
6) when round number input is letter of string
7) when round number input is empty
8) when round number input is less than 1

Hogu59 committed:
test: test for printing winner

Test conducted to check getting right winner.

Hogu59 committed:
feat(dependency): use junit5 explicitly

Modification of codes to use junit5 explicitly.

Hogu59 committed:
docs: modify readme

All requirements were checked, including exceptions.

Hogu59 committed:
refactor(controller): seperate constants

Move the values of hard-coding to companion object.
To make more efficient to re-use.

Hogu59 committed:
refactor(model): seperate constants

Move the values of hard-coding to companion object.
To make more efficient for further modification.

Hogu59 committed:
refactor(controller): seperate functions

Function seperated to make depth shallower.

Hogu59 committed:
refactor(controller): modify error message

Error message divided by each cases.

Hogu59 committed:
refactor(model): modify error message

Error message divided by each cases.

Hogu59 committed:
refactor(controller): make functions private

Functions only used in local are changed private.

Hogu59 committed:
refactor(model): make functions private

Functions only used in local are changed private.

KwonDae commented on README.md:
기능 목록을 정말 잘 작성해주셨네요! 💯
커밋도 기능 목록 단위로 작성해보면 어떨까요?
커밋 횟수가 많은 커밋이 효율적일지 한번 고민해보시기 바랍니다 🧐

Hogu59 replied:
기능목록에 대해서 좋게 평가해주셔서 감사합니다.

커밋 횟수에 있어서 조금 고민되는 부분이 있었는데, 프로젝트 단위에 비해 너무 세분화해서 커밋했던 것 같네요...ㅎㅎ
조금 더 묶여질 수 있는 단위로 묶어서 커밋하는 것에 대해서 연습해보도록 하겠습니다!

KwonDae commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,152 @@
+        require(names.size >= MIN_NAMES_SIZE) { ERROR_CAR_LESS_THAN_TWO }
+        require(names.size <= MAX_NAMES_SIZE) { ERROR_CAR_GREATER_THAN_TWENTY }
+        require(names.distinct().size == names.size) { ERROR_NAME_DUPLICATION }
+    }
```
> 입력값 검증(Validation)을 InputView 에서 하는게 좋은지, 아니면 model 혹은 controller 에서 진행하는 것이 좋은지 알고 싶습니다.

어느 곳에서 검증하는게 좋을지에 대해서 고민이 많으셨던거 같아요.
검증을 담당하는 클래스를 두어서 책임을 분리하는것으로 시작해보는건 어떨까요? 그렇다면 model, controller 어디서든 validation   담당자를 호출해 사용이 가능하겠네요!

> (이번 미션에서는 model과 controller에서 유효성 검증을 했는데, depth가 깊어지는 문제가 있었습니다... 혹시 depth를 줄이는 방법이 있다면 알 수 있을까요..?)

가장 간단한 방법은 함수분리 입니다.
모든 검증을 한 뒤에(여기서 depth가 길어질 수 있겠죠) 움직이라는 명령을 내릴 수 있겠지만,

객체가 자동차를 받아, 자동차에게 단지 메세지만 던져 자동차 스스로 판단해 움직일지 말지 역할을 부여하는것 또한 좋은 방법입니다.
이런식으로 역할을 분리해서 depth를 줄일 수 있습니다.

> 함수(또는 메서드)의 길이가 15라인을 넘어가지 않도록 구현한다.
> 함수(또는 메소드)가 한 가지 일만 하도록 최대한 작게 만들어라.

Hogu59 replied:
Validation 클래스에 대해 말씀해주셔서 한번 클래스를 추가로 구현해보았습니다.
확실히 유효성 검증에 대한 부분을 분리하니 controller에서 처리하는 내용이 줄어드는 것 같았습니다.
이전에는 controller에서 모든 것을 처리하려고 하다보니 controller만 너무 비대해지는 것은 아닌지 걱정이 되었는데 피드백 덕분에 조금 해결방법을 찾은 것 같습니다!

KwonDae commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,152 @@
+                validateNames(names)
+                names.forEach { name ->
+                    Car(name)
+                }
```
validation 하는 함수가 검증할 대상(target)을 받아,
이상이 있으면 throw,
정상이면 그 값을 return 하여 depth를 줄일 수 있겠네요.

먼저 validation 하는 로직부터 분리해보고 하나씩 하나씩 제거해볼까요?

Hogu59 replied:
validation 로직을 분리해보았습니다! Validation 클래스로 분리하여 작성해보았는데 한 번 확인해주시면 감사하겠습니다!

KwonDae commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,152 @@
+        showWinners()
+    }
+
+    private fun getNames(): List<String> {
+        outputView.printCarNamesGuide()
+        lateinit var names: List<String>
```
> Controller로 Race class 를 사용하고 있습니다. 이때, getNames() 메서드에서 OutputView를 통해 입력가이드문을 출력하는 부분이 있는데, 이러한 출력부를 별도의 메서드로 분리하는 것이 좋은지, 아니면 메소드명을 수정하는 것이 좋은지 알고 싶습니다!

가장 중요한건 책임 분리 & 함수 사용 입니다.

```
fun getCars : List<Car> {
   OutputView.printCarNamesGuide()
   val names = InputValidator.validationName(InputView.InputName)
   return makeCars(names)
}

```

요런 3줄로 책임 분리, 함수를 통해 입출력, 검증, 자동차 생성이 가능하겠죠?

KwonDae commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    companion object {
+        const val ERROR_REQUEST_INPUT = "[Error] 값을 입력해 주세요"
+        const val ERROR_CAR_LESS_THAN_TWO = "[Error] 자동차 이름은 2대 이상 입력해 주세요."
+        const val ERROR_CAR_GREATER_THAN_TWENTY = "[Error] 자동차 이름은 20대 이하로 입력해 주세요."
+        const val ERROR_NAME_DUPLICATION = "[Error] 중복된 자동차 이름은 작성할 수 없습니다."
+        const val ERROR_ROUND_LESS_THAN_ONE = "[Error] 1회 이상 입력해 주세요."
+        const val ERROR_INPUT_NUMBER_ONLY = "[Error] 숫자만 입력해 주세요."
+
+        const val COMMA = ","
+
+        const val MIN_NAMES_SIZE = 2
+        const val MAX_NAMES_SIZE = 20
+        const val MIN_ROUND_NUMBER = 1
+        const val INITIAL_ROUND_NUMBER = 0
```
상수 활용을 잘 해주셨네요 💯
다른 곳에서 쓰이지 않는 상수라면 한정자를 통해 제한해볼까요?

Hogu59 replied:
꼼꼼한 피드백 감사합니다!
companion object 내부의 상수에 대해서 한정자를 사용하는 것에 대해서는 간과하고 있었는데 덕분에 알게 되었습니다!

KwonDae commented on src/main/kotlin/racingcar/model/Car.kt:
```diff
@@ -0,0 +1,32 @@
+
+import kotlin.random.Random
+
+data class Car(val name: String, var position: Int = INITIAL_POSITION) {
```
data class를 잘 활용해주셨지만, constructor에서 var 으로 선언하는것에 대한 부분을 생각해볼까요?
변경이 되어야 하는 변수이기에 var을 사용했지만 [링크](https://stackoverflow.com/questions/49688571/can-the-parameter-in-data-class-be-var-in-kotlin)를 참고해보시고 어떤 문제가 있을 수 있을지 생각해보세요

정리하면, data class 를 사용하는 목적이 무엇이고, 생성자에서 선언하는것과 [body 에서 선언](https://kotlinlang.org/docs/data-classes.html#properties-declared-in-the-class-body)하는것이 어떤 차이점이 있을지에 대해서 고민해보세요 👍

Hogu59 replied:
피드백을 통해 data class 에 대해서 조금 더 알게 된 것 같습니다.
data class의 immutability 를 위해서 val 로 선언하는 것이 (불가피한 상황을 제외한다면) 바람직하다는 것을 처음 알게 되었습니다.

이 코드를 작성할 때만 해도 class 와 data class의 차이를 정확히 몰라 그냥 사용했었는데, 좀 더 자세히 공부해서 목적에 맞는 사용을 지향해보도록 하겠습니다!

KwonDae commented on src/main/kotlin/racingcar/model/Car.kt:
```diff
@@ -0,0 +1,32 @@
+
+data class Car(val name: String, var position: Int = INITIAL_POSITION) {
+    init {
+        require(name.isNotEmpty()) { ERROR_CAR_NAME_BLANK }
+        require(name.length in MIN_NAME_LENGTH..MAX_NAME_LENGTH) { ERROR_CAR_NAME_OUT_OF_RANGE }
+        require(!name.contains(SPACE)) { ERROR_CAR_NAME_CONTAINS_SPACE }
```
require 활용!! 👍

KwonDae commented on src/test/kotlin/racingcar/RaceTest.kt:
```diff
@@ -0,0 +1,125 @@
+                race.validateNames(emptyList())
+            }
+
+        assertEquals(
```
assertEquals 대신 assertThat 을 활용해보는건 어떨까요?
아래 링크를 참고해보세요!
https://jongmin92.github.io/2020/03/31/Java/use-assertthat/

Hogu59 replied:
첨부해주신 링크로 많은 도움이 되었습니다!
또한 비슷한 내용으로 검색했을 때에도 assertThat을 사용하는것이 더 가독성이 높아지거나 유용한 부분들이 있다는 것을 알게 되었습니다!
참고하여 추후 테스트 코드 작성시에도 반영해보도록 하겠습니다.

KwonDae reviewed (changes requested):
안녕하세요!

이번 자동차 미션을 함께하게된 제이든입니다.
구현을 깔끔하게 잘 해주셨어요!
질문주신 부분인 Race.kt 부분을 중점으로 코멘트 드린 내용들 확인해주시고
피드백 반영 후 다시 리뷰 요청 부탁드립니다 🙇‍♂️
화이팅 👍

Hogu59 committed:
fix(test): change error msg

Add constants error messages as companion objects.
Change comparing messages for tests.

Hogu59 committed:
fix(controller): fix problem(scope of cars)

Error was occured because of the scope issue.
1) getWinners function get cars as an argument.
2) judgeWinners function uses the lateinit member variable.

To fix the problem, cars(List<Car>) parameter was added.
But, it should be considered again (using three parameter is not recommended.

Hogu59 committed:
docs: readme added(need to be fixed)

Readme file modified. Something need to be fixed.

Hogu59 committed:
docs: readme added(need to be fixed)

Readme file modified. Something need to be fixed.

Hogu59 committed:
refactor(test): modify assert statements

Assert statements modified.
assertEquals -> assertThat

Hogu59 committed:
refactor(view, model): Erase companion objects.

Erase unnecessary companion objects.

Hogu59 committed:
refactor(validation): Seperate validation parts

Make Validation class to seperate validation parts.

Hogu59 committed:
docs: check the fixed list

Readme file modified. Fixed lists were checked.

Hogu59 committed:
refactor(model): Change Car to class

Change Model(Car) to (just) class from data class

KwonDae commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,125 @@
+        lateinit var names: List<String>
+
+        while (true) {
+            kotlin.runCatching {
+                names = inputView.readCarNames().split(COMMA)
+                validateNames(names)
+            }
+                .onSuccess { return names }
+                .onFailure { exception -> println(exception.message) }
```
runCatching 활용 👍

KwonDae commented on src/main/kotlin/racingcar/controller/Race.kt:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    fun validateNames(names: List<String>) {
+        Validation().carNameList(names)
+        validateEachName(names)
```
입력받은 `Name List`와 `각 Name `개별 검증을 분리한 이유가 있을까요?

Hogu59 replied:
유효성 검증을 두 부분으로 나눠서 하면 함수를 각 기능별로 분리할 수 있을것 같아서 그렇게 했었습니다!
두 부분에 있어서 발생 가능한 예외상황이 다르다고 생각해서 전체 입력인 이름 목록과 각 이름에 대한 유효성 검증을 나누었습니다!

`nameList`의 경우 입력된 이름 목록 자체가 유효한지 검증하려고 했습니다.
 - 예상 오류
   1) 입력 자체가 공백 ("")
   2) 이름이 하나만 있는 경우(차가 한대이면 레이싱 경주가 불가
   3) 자동차 대수 확인을 통해 이름 목록 크기가 2 미만 20초과에 대해서는 예외처리
   4) 중복값 확인 등

`name`의 경우에 대해서는 각 이름이 유효성을 검증하려고 했습니다.
- 예상 오류
  1) 이름이 공백 (ex. "pobi", "", "crong")  
  2) 이름이 5자 초과 (ex. "crongcrong")
  3) 이름에 공백이 있는 경우 등(ex. "po bi")

KwonDae commented on src/main/kotlin/racingcar/model/Car.kt:
```diff
@@ -0,0 +1,23 @@
+import kotlin.random.Random
+
+class Car(val name: String, var position: Int = INITIAL_POSITION) {
+    init {
+        Validation().carName(name)
+    }
```
Car 객체를 생성할때는 constructor로 들어오는 값들은 이미 검증이 완료된 상태로 들어오게 하는건 어떨까요?
Car 와 Validation 간 의존성은 없어도 됩니다.

Hogu59 replied:
알려주셔서 감사합니다! 말씀해주신대로 이 프로젝트에서는 이미 검증된 값이 넘어오면 문제가 없을것 같습니다!
해당 코드 수정하였습니다, 확인해주시면 감사하겠습니다!

Q)
큰 프로젝트에서는 다양한 예외상황들이 존재할 것이라고 생각하는데, 그런 경우에 있어서는 유효성 검증을 중복으로도 하는 경우가 있는지 궁금합니다! (ex. InputView에서 입력값을 1차 검증 -> Controller 에서 2차 검증)

Q)
실제로 유효성 검증이 중복으로도 사용된다면 보통 어디에서 이루어지는지 궁금합니다! (ex, view-controller, controller-model, view-controller-model 등등.)

KwonDae commented on src/main/kotlin/racingcar/validator/Validation.kt:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    companion object {
+        private const val ERROR_REQUEST_INPUT = "[Error] 값을 입력해 주세요"
+        private const val ERROR_CAR_LESS_THAN_TWO = "[Error] 자동차 이름은 2대 이상 입력해 주세요."
+        private const val ERROR_CAR_GREATER_THAN_TWENTY = "[Error] 자동차 이름은 20대 이하로 입력해 주세요."
+        private const val ERROR_NAME_DUPLICATION = "[Error] 중복된 자동차 이름은 작성할 수 없습니다."
+        private const val ERROR_ROUND_LESS_THAN_ONE = "[Error] 1회 이상 입력해 주세요."
+        private const val ERROR_INPUT_NUMBER_ONLY = "[Error] 숫자만 입력해 주세요."
+
+        private const val ERROR_CAR_NAME_BLANK = "[Error] 자동차 이름은 공란일 수 없습니다."
+        const val ERROR_CAR_NAME_OUT_OF_RANGE = "[Error] 자동차 이름은 1자 이상 5자 이내여야 합니다."
+        private const val ERROR_CAR_NAME_CONTAINS_SPACE = "[Error] 이름에 공백이 없어야 합니다."
+
+        private const val SPACE = " "
+
+        private const val MIN_NAME_LENGTH = 1
+        private const val MAX_NAME_LENGTH = 5
+        private const val MIN_NAMES_SIZE = 2
+        private const val MAX_NAMES_SIZE = 20
+        private const val MIN_ROUND_NUMBER = 1
+    }
```
💯

KwonDae reviewed (approved):
피드백 반영해주시느라 고생 많으셨습니다!
스스로 고민하시고 객체를 객체스럽게 사용하시려는 변화가 보여 인상깊었습니다.

남겨드린 피드백은 다음 미션때 함께 반영해주세요
2단계 미션도 화이팅입니다💪
